rfcs/2021-06-15-shipping-blade.md

```diff
@@ -0,0 +1,438 @@
+
+### `main`, `module`, `browser`, `react-native` fields in `package.json`
+
+The older way to define this consumption information used to be via `main`, `module`, `browser`, `react-native` fields in `package.json`.
+* `main`: defines entry point for commonjs bundle
+* `module`: defines entry point for esm bundle.
+* `browser`: defines entry point for browser specific bundle. If not provided, then fallback to `main` or `module`.
+* `react-native`: defines entry point for react-native bundle. If not provided, then fallback to `main`.
+* The problem with all of the above fields is that they accept a single string value and can be useful if you have single bundle of your package. For our use case we have multiple bundles. Hence, this won't be useful for us.
+
+### Conclusion
+Now based on the above discussions we definitely want to go ahead with `exports` map to define subpaths for our package but that is not supported by all the tools. So what do we do? We need to provide fallback for backward compatibility. So how do we do that? We generate `components`, `tokens`, `utils`  files at the root and re-export things from `build`.
+```
+blade/
+  build/
+    components/
+      index.web.js
+      index.native.js
+    tokens/
+      index.web.js
+      index.native.js
+    utils/
+      index.web.js
+      index.native.js
+  src/
+  components.js
+    export * from './build/components';
+  tokens.js
+    export * from './build/tokens';
+  utils.js
+    export * from './build/utils';
+```
+
+With the above we would still be able to import in following way:
+```js
+import {ThemeProvider, useTheme} from '@razorpay/blade/components';
+import {paymentTheme, globalColors} from '@razorpay/blade/tokens';
+```
+
+# Drawbacks/Constraints
+I've already spoken about multiple approaches and constraints of each of them in the respective sections itself.
+
+# Alternatives
+I've already spoken about multiple approaches and constraints of each of them in the respective sections itself.
+
+# Adoption strategy
+There's no adoption strategy as such. We'll just need to document the usage of design system once it's published.
+
+# Open Questions
+1. Do we need any changes in our src directory structure to support our needs?
+2. We have custom types defined at a lot of place so how do we merge them with the types generated by tsconfig?
+   * Maybe use something similar to [this library](https://www.npmjs.com/package/npm-dts).
+
+# References
+* [NodeJS packages support](https://nodejs.org/api/packages.html)
+* [`exports` map](https://docs.skypack.dev/package-authors/package-checks#export-map)
+* [Webpack `exports` field handling and examples](https://webpack.js.org/guides/package-exports/)
+* [`exports` implementation PR in webpack](https://github.com/webpack/webpack/pull/10953)
+* [`exports` field issue in metro](https://github.com/facebook/metro/issues/670)
```

Comment:
Can you also add some references around how other popular open-source DS are doing this?

Reply:
No one has added it as a post. I had to dig through their codebases and all of them have mixed strategies.